product/util.py

```diff
@@ -171,164 +211,197 @@ def diff_products(left_product, right_product, ignore_component_names=()):
+        )
+    if type(right_component) is not cm.Component:
+        raise NotImplementedError(
+            f'unsupported {type(right_component)=}. Only cm.Component is supported',
+        )
 
-    left_names_to_imgs = {i.name(): i for i in left_images}
-    right_names_to_imgs = {i.name(): i for i in right_images}
+    left_resource_identities = {
+        r.identity(left_component.resources + right_component.resources): r
+        for r in left_component.resources
+    }
+    right_resource_identities_to_resource = {
+        r.identity(left_component.resources + right_component.resources): r
+        for r in right_component.resources
+    }
 
-    img_diff = ImageDiff(
+    resource_diff = ResourceDiff(
         left_component=left_component,
         right_component=right_component,
     )
 
-    if left_images == right_images:
-        return img_diff
-
-    for name, img in left_names_to_imgs.items():
-        if not name in right_names_to_imgs:
-            img_diff.irefs_only_left.add(img)
+    if left_resource_identities.keys() == right_resource_identities_to_resource.keys():
+        return resource_diff
+
+    left_names_to_resource = {r.name: r for r in left_component.resources}
+    right_names_to_resource = {r.name: r for r in right_component.resources}
+    # get left exclisive resources
+    for resource in left_resource_identities.values():
+        if not resource.name in right_names_to_resource:
+            _add_if_not_duplicate(resource_diff.resource_refs_only_left, resource)
+
+    # get right exclusive images
+    for resource in right_resource_identities_to_resource.values():
+        if not resource.name in left_names_to_resource:
+            _add_if_not_duplicate(resource_diff.resource_refs_only_right, resource)
+
+    def enumerate_group_pairs(
+        left_resources: typing.List[cm.Resource],
+        right_resources: typing.List[cm.Resource]
+    ) -> typing.Tuple[typing.List[cm.Resource], typing.List[cm.Resource]]:
+        # group the images with the same name on both sides
+        for key in left_names_to_resource.keys():
+            left_resource_group = [r for r in left_resources if r.name == key]
+            right_resource_group = [r for r in right_resources if r.name == key]
+
+            # key is always in left group
+            if len(right_resource_group) == 0:
+                continue
+            else:
+                yield (left_resource_group, right_resource_group)
 
-    for name, img in right_names_to_imgs.items():
-        if not name in left_names_to_imgs:
-            img_diff.irefs_only_right.add(img)
+    for left_resource_group, right_resource_group in enumerate_group_pairs(
+        left_resources=left_component.resources,
+        right_resources=right_component.resources,
+    ):
+        if len(left_resource_group) == 1 and len(right_resource_group) == 1:
+            # if versions are equal resource will be ignored, resource is unchanged
```

Comment:
actually, this really only holds true if the identities be equal. however, let us leave it like this for now, to not make things unnecessarily complicated..